### PR TITLE
[SYCL] Implement forward_progress extension API

### PR DESCRIPTION
**Pull request by lbushi25**
This PR implements the API for the forward progress guarantees extension defined here: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc
~~It is complete except that it does not verify that the forward progress guarantees requested by a kernel submission are actually supported by the device to which the kernel is submitted. That will be done in a later PR.~~

### CODE REVIEW

File: sycl/source/detail/device_info.hpp
Comment (steffenlarsen) [inline sycl/source/detail/device_info.hpp:2287]: Seems like a good chunk of this implementation is repeated. Could we move some of it to a separate function that we can reuse?

File: sycl/source/detail/device_info.hpp
Comment (steffenlarsen) [inline sycl/source/detail/device_info.hpp:2281]:
```suggestion
    int guarantee_val = static_cast<int>(guarantee);
    res.reserve(forward_progress_guarantee_size - guarantee_val);
```

File: sycl/source/detail/device_info.hpp
Comment (steffenlarsen) [inline sycl/source/detail/device_info.hpp:2278]: The file is a little inconsistent with the coding style, but unless the file has a consistent coding style, we should use the [LLVM Coding Standard](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly) mandated naming. This does not apply to public interfaces.

**Comment by gmlueck** [comment]
> It is complete except that it does not verify that the forward progress guarantees requested by a kernel submission are actually supported by the device to which the kernel is submitted. That will be done in a later PR.

Can you elaborate on what part of the spec is not yet implemented?  Is it just this statement about throwing an exception that is not yet implemented?

> New kernel properties are introduced to allow developers to declare that a given kernel requires specific forward progress guarantees for correctness. If a kernel is submitted to a device that cannot satisfy the request for specific progress guarantees, the implementation must throw an `exception` with the `errc::feature_not_supported` error code.

Is there an ETA for when the remaining parts will be implemented?  I'm wondering when we want to declare the extension "implemented" and move the extension document to the "experimental" directory.

Also, does this PR also implement [sycl_ext_oneapi_launch_queries](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_launch_queries.asciidoc)?  Those queries are required in order to correctly use the forward progress guarantee extension.

**Comment by lbushi25** [comment]
> > It is complete except that it does not verify that the forward progress guarantees requested by a kernel submission are actually supported by the device to which the kernel is submitted. That will be done in a later PR.
> 
> Can you elaborate on what part of the spec is not yet implemented? Is it just this statement about throwing an exception that is not yet implemented?
> 
> > New kernel properties are introduced to allow developers to declare that a given kernel requires specific forward progress guarantees for correctness. If a kernel is submitted to a device that cannot satisfy the request for specific progress guarantees, the implementation must throw an `exception` with the `errc::feature_not_supported` error code.
> 
> Is there an ETA for when the remaining parts will be implemented? I'm wondering when we want to declare the extension "implemented" and move the extension document to the "experimental" directory.
> 
> Also, does this PR also implement [sycl_ext_oneapi_launch_queries](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_launch_queries.asciidoc)? Those queries are required in order to correctly use the forward progress guarantee extension.

Hi Greg,

Yes, it is the statement about throwing an exception when the progress guarantee cannot be met that is not yet implemented. As for the ETA, I'm working on it in parallel with this PR but decided to not include it in this PR because it looks to me like it will require some refactoring. I will ping you soon about guidance on this.

Finally, no, this PR does not at the moment implement the sycl_ext_oneapi_launch_queries extension but now that you brought it to my attention, I can put it in the next PR along with the verification of kernel properties.

**Comment by gmlueck** [comment]
> Yes, it is the statement about throwing an exception when the progress guarantee cannot be met that is not yet implemented. As for the ETA, I'm working on it in parallel with this PR but decided to not include it in this PR because it looks to me like it will require some refactoring. I will ping you soon about guidance on this.
> 
> Finally, no, this PR does not at the moment implement the sycl_ext_oneapi_launch_queries extension but now that you brought it to my attention, I can put it in the next PR along with the verification of kernel properties. I will probably only implement the necessary parts of that extension needed to query progress guarantees so as to not stray too much from the original objective that this PR was created for.

Thanks for the update.  Let's wait for the next PR before moving the extension document to "experimental".

File: sycl/test/extensions/forward_progress_L0_gpu.cpp
Comment (Pennycook) [inline sycl/test/extensions/forward_progress_L0_gpu.cpp:35]: Nit: I found it a little difficult to read through this (and other tests) to check that we hadn't missed any important cases.  I think it would be helpful to group things by execution scope and break them apart:

```c++
  // Check guarantees for execution_scope::root_group
  auto guarantees = d.get_info<info::device::work_group_progress_capabilities<
      execution_scope::root_group>>();
  assert(guarantees == ref2);

  guarantees = d.get_info<info::device::sub_group_progress_capabilities<
      execution_scope::root_group>>();
  assert(guarantees == ref2);

  guarantees = d.get_info<info::device::work_item_progress_capabilities<
      execution_scope::root_group>>();
  assert(guarantees == ref1);

  // Check guarantees for execution_scope::work_group
  guarantees = d.get_info<info::device::sub_group_progress_capabilities<
      execution_scope::work_group>>();
  assert(guarantees == ref2);

  guarantees = d.get_info<info::device::work_item_progress_capabilities<
      execution_scope::work_group>>();
  assert(guarantees == ref1);

  // Check guarantees for execution_scope::sub_group
  guarantees = d.get_info<info::device::work_item_progress_capabilities<
      execution_scope::sub_group>>();
  assert(guarantees == ref1);
```

**Comment by lbushi25** [comment]
I have made the necessary changes to allow the kernel to query what progress guarantees it can request on a given device through the ext_oneapi_get_info function. Relevant changes are in kernel_launch_queries.hpp and kernel.hpp. PI codes for these kernel queries have also been added in pi.h.

**Comment by lbushi25** [comment]
I've refactored the device_impl.{hpp, cpp} files and implemented verification of the forward progress requirements in the handler.{hpp, cpp}.

File: sycl/source/handler.cpp
Comment (Pennycook) [inline sycl/source/handler.cpp:1426]: I'm not sure what our convention is here, but I think "this device" looks better than "your device!".  Same comment for the other two exceptions.

```suggestion
          "supported by this device.");
```

File: sycl/include/sycl/handler.hpp
Comment (Pennycook) [inline sycl/include/sycl/handler.hpp:1000]: I'm surprised that we don't appear to be modifying this (or something like it) based on the progress properties.  This might be my fault for not documenting things clearly in the [implementation notes](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc#implementation-notes) section of the specification.

Although the table there says "concurrent" for "GPU & Level Zero", that's because the table was intended to represent the _maximum_ capability, not the default.  By default, I think work-group progress is "parallel" -- if the user requested "concurrent" then we need a cooperative launch.  I'm honestly not sure if a cooperative launch is required for "concurrent" at sub-group scope.

Enabling cooperative launch for "concurrent" at work-group or sub-group scope might be a good idea for now (to make sure that users get the guarantee they asked for) but I think there needs to be some follow-up design work to decide how to express this at a UR level.  Different backends may behave differently, and may not require a cooperative launch in the cases that Level Zero does.

Comment (lbushi25) [reply sycl/include/sycl/handler.hpp:1000]: > I'm surprised that we don't appear to be modifying this (or something like it) based on the progress properties. This might be my fault for not documenting things clearly in the [implementation notes](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc#implementation-notes) section of the specification.
> 
> Although the table there says "concurrent" for "GPU & Level Zero", that's because the table was intended to represent the _maximum_ capability, not the default. By default, I think work-group progress is "parallel" -- if the user requested "concurrent" then we need a cooperative launch. I'm honestly not sure if a cooperative launch is required for "concurrent" at sub-group scope.
> 
> Enabling cooperative launch for "concurrent" at work-group or sub-group scope might be a good idea for now (to make sure that users get the guarantee they asked for) but I think there needs to be some follow-up design work to decide how to express this at a UR level. Different backends may behave differently, and may not require a cooperative launch in the cases that Level Zero does.

Thanks for the knowledge shared, I'll make the change and add a TODO item as a comment asking to clarify this at a later stage.

Comment (Pennycook) [reply sycl/include/sycl/handler.hpp:999]: @JackAKirk - Does the same logic here apply to the CUDA backend?  I think it does, but I just want to double check.  If the same logic applies, would enabling a cooperative kernel call for CUDA here work as expected, or should the CUDA backend work be left for a future PR?

Comment (lbushi25) [reply sycl/include/sycl/handler.hpp:999]: > @JackAKirk - Does the same logic here apply to the CUDA backend? I think it does, but I just want to double check. If the same logic applies, would enabling a cooperative kernel call for CUDA here work as expected, or should the CUDA backend work be left for a future PR?

@JackAKirk ping

Comment (JackAKirk) [reply sycl/include/sycl/handler.hpp:999]: Sorry I missed the first ping.

At the subgroup level I read that nvidia post volta satisfies the C++ forward progress guarantees which seems to coincide with your extension document. The best resource I know of explaining this is https://developer.nvidia.com/blog/inside-volta/ 
I don't have access to it but this doc might provide further details: https://developer.nvidia.com/blog/inside-volta/
Also see https://on-demand.gputechconf.com/gtc/2018/presentation/s81006-volta-architecture-and-performance-optimization.pdf

Now wrt the question I think you are asking: 
- how does nvidia `cudaLaunchCooperativeKernel` "weaken" these guarantees as defined in your extension?

 I don't know the answer to this and I don't see mention of it in Nvidia docs etc.
I don't know the details of C++ iso forward progress guarantees but I think I get the basic idea; the other thing that is relevant to note is that `cudaLaunchCooperativeKernel` is not the only way that these guarantees might be affected from what I understand. `cudaLaunchKernelExC`

https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION_1ge236ecdbbaf7cf47a806bba71c1d03c4

 is a newer api (requires sm90 "hopper") and allows access of distributed shared memory between different thread blocks, as well as interthread block atomics to control such distributed shared memory that requires their ptx 7.8 memory model:

https://dl.acm.org/doi/pdf/10.1145/3470496.3533045

It is as far as I can see completely separate from `cudaLaunchCooperativeKernel` and provides non overlapping functionality. The impression I get is that going forward this will be a more important means of inter-thread block interaction than `cudaLaunchCooperativeKernel` (I note also that the multi device version of this api has been deprecated).

Now I may have misunderstood some things but I guess that both `cudaLaunchCooperativeKernel` and `cudaLaunchKernelExC` (with associated cluster_group functionality) https://github.com/intel/llvm/pull/13594 might affect your guarantees, and possibly in different ways.

Hope this is in some way useful.

Comment (lbushi25) [reply sycl/include/sycl/handler.hpp:999]: > Sorry I missed the first ping.
> 
> At the subgroup level I read that nvidia post volta satisfies the C++ forward progress guarantees which seems to coincide with your extension document. The best resource I know of explaining this is https://developer.nvidia.com/blog/inside-volta/ I don't have access to it but this doc might provide further details: https://developer.nvidia.com/blog/inside-volta/ Also see https://on-demand.gputechconf.com/gtc/2018/presentation/s81006-volta-architecture-and-performance-optimization.pdf
> 
> Now wrt the question I think you are asking:
> 
> * how does nvidia `cudaLaunchCooperativeKernel` "weaken" these guarantees as defined in your extension?
> 
> I don't know the answer to this and I don't see mention of it in Nvidia docs etc. I don't know the details of C++ iso forward progress guarantees but I think I get the basic idea; the other thing that is relevant to note is that `cudaLaunchCooperativeKernel` is not the only way that these guarantees might be affected from what I understand. `cudaLaunchKernelExC`
> 
> https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION_1ge236ecdbbaf7cf47a806bba71c1d03c4
> 
> is a newer api (requires sm90 "hopper") and allows access of distributed shared memory between different thread blocks, as well as interthread block atomics to control such distributed shared memory that requires their ptx 7.8 memory model:
> 
> https://dl.acm.org/doi/pdf/10.1145/3470496.3533045
> 
> It is as far as I can see completely separate from `cudaLaunchCooperativeKernel` and provides non overlapping functionality. The impression I get is that going forward this will be a more important means of inter-thread block interaction than `cudaLaunchCooperativeKernel` (I note also that the multi device version of this api has been deprecated).
> 
> Now I may have misunderstood some things but I guess that both `cudaLaunchCooperativeKernel` and `cudaLaunchKernelExC` (with associated cluster_group functionality) #13594 might affect your guarantees, and possibly in different ways.
> 
> Hope this is in some way useful.

Thank you @JackAKirk !
@Pennycook Should I leave the cuda changes for a future PR?

Comment (Pennycook) [reply sycl/include/sycl/handler.hpp:999]: Thanks, @JackAKirk .  This is useful -- I didn't know about the existence of `cudaLaunchKernelExc`.

@lbushi25 - My recommendation here for now is to change the implementation to:

```c++
    if (Guarantee == forward_progress::concurrent)
      setKernelIsCooperative(true);
```

...so that it's required for all backends.  It sounds like the UR will then need to decide between `cudaLaunchCooperativeKernel`, `cudaLaunchKernelExC` for CUDA.

I think moving the complexity out of the headers and into UR is the right thing to do here.

**Review by Pennycook — APPROVED** [review]

**Comment by lbushi25** [comment]
@intel/llvm-reviewers-runtime ping for review

**Comment by lbushi25** [comment]
@intel/llvm-reviewers-runtime second ping as all pre-commit tests are now passing.

File: sycl/source/detail/device_impl.cpp
Comment (KseniyaTikhomirova) [inline sycl/source/detail/device_impl.cpp:875]: Host device was part of SYCL 1.2 but was removed in SYCL2020. 
We plan to remove any signs of it in RT source code soon.  Would be glad to not introduce new code for it.

tagging @aelovikov-intel for awareness.

File: sycl/source/detail/device_info.hpp
Comment (KseniyaTikhomirova) [inline sycl/source/detail/device_info.hpp:2228]: same for device info host. aspect::host is deprecated https://github.com/intel/llvm/blob/a43ff24492fdb9ced7f2c906a593f20904d2a261/sycl/include/sycl/info/aspects_deprecated.def#L5
will it not compile now if we just not add get_device_info_host?

File: sycl/source/handler.cpp
Comment (KseniyaTikhomirova) [inline sycl/source/handler.cpp:1469]: to align naming with other parameters

File: sycl/source/detail/device_info.hpp
Comment (lbushi25) [reply sycl/source/detail/device_info.hpp:2228]: Yeah, unfortunately I was getting compilation errors when I had not defined get_device_info_host.

Comment (KseniyaTikhomirova) [reply sycl/source/detail/device_info.hpp:2228]: I see, ok then.

File: sycl/test/extensions/forward_progress_L0_gpu.cpp
Comment (AlexeySachkov) [inline sycl/test/extensions/forward_progress_L0_gpu.cpp:1]: Any tests which require actual HW should be in `e2e` folder. You should also use `%{build}` and `%{run}` substitutions in there

**Review by AlexeySachkov — CHANGES_REQUESTED** [review]

File: sycl/source/handler.cpp
Comment (KseniyaTikhomirova) [reply sycl/source/handler.cpp:1469]: it is just a NIT but idea was to do just the opposite. see other functions (e.g. line 1457).

**Review by KseniyaTikhomirova — APPROVED** [review]
LGTM

**Comment by lbushi25** [comment]
@intel/llvm-gatekeepers This is ready for merge.

**Comment by againull** [comment]
> @intel/llvm-gatekeepers This is ready for merge.

Merge button is blocked because of requested changes,  explicit approval from @AlexeySachkov  is needed.

**Review by AlexeySachkov — APPROVED** [review]

**Comment by lbushi25** [comment]
@intel/llvm-gatekeepers The PR is now approved. The AWS failures are unrelated to my PR.

**Comment by lbushi25** [comment]
@intel/llvm-gatekeepers ping

File: sycl/test/include_deps/sycl_accessor.hpp.cpp
Comment (aelovikov-intel) [inline sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: Why is necessary?

Comment (lbushi25) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: info/info_desc.hpp needs info about forward progress enums so it includes forward_progress.hpp which means it will appear as a transitive dependecy.

Comment (aelovikov-intel) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: Are you saying that the entire extensions is necessary as part of "core" SYCL include? I can't agree with that, because then our compile time will be skyrocketing going forward with any new PR and it's already terrible.

Comment (lbushi25) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: I don't know whether it necessary as part of "core" SYCL but I do know that it is necessary in the files where I have included it, at least to the best of my knowledge. During pre-commit testing, some tests were failing because they were expecting to see my header file as a (direct or indirect) dependency for certain headers. So I added these check lines to make these tests pass. Perhaps this points to a larger systemic issue with the way we've designed our tests/code but that will need to be handled separately in my opinion.

File: sycl/include/sycl/handler.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/handler.hpp:951]: I'm not sure how having this not inlined is helping in readability. You have to duplicate all the `constexpr if` from the caller, I don't even see what common code this helper abstracts away.

File: sycl/test/include_deps/sycl_accessor.hpp.cpp
Comment (aelovikov-intel) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: >  but that will need to be handled separately in my opinion.

How do you suggest we find a person who would be refactoring the entire SYCL implementation if all other people would be contributing to the constantly growing compile time using that excuse?

Comment (gmlueck) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: Just noticed this conversation when I saw f783279655030d27ae94e9a510152c96c221d94b.  The forward progress extension does not document that the application must include a special header file to use it, therefore, the assumption is that you only need to include `<sycl/sycl.hpp>`.  This is the same for all of our other extensions.  I don't think this extension should be different from all the others.

I agree that we should have a more granular set of headers in SYCL.  I think the right way to do this is to keep `<sycl/sycl.hpp>` as the "uber" header that includes everything, and then we can also create a set of more granular headers that include specific features.  We can also create separate headers for each extension (to the extent possible) as part of this effort.

Comment (lbushi25) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: Thanks for the clarification @gmlueck !

Comment (lbushi25) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: @aelovikov-intel 

> Just noticed this conversation when I saw [f783279](https://github.com/intel/llvm/commit/f783279655030d27ae94e9a510152c96c221d94b). The forward progress extension does not document that the application must include a special header file to use it, therefore, the assumption is that you only need to include `<sycl/sycl.hpp>`. This is the same for all of our other extensions. I don't think this extension should be different from all the others.
> 
> I agree that we should have a more granular set of headers in SYCL. I think the right way to do this is to keep `<sycl/sycl.hpp>` as the "uber" header that includes everything, and then we can also create a set of more granular headers that include specific features. We can also create separate headers for each extension (to the extent possible) as part of this effort.

The second paragraph of the quoted reply summarizes where I was going at when I said that we might have a more systemic issue. For now, however, I'm just respecting the status quo as previous extensions have done. In each of the files that forward_progress.hpp is included, the inclusion is necessary and compilation fails otherwise.

Comment (aelovikov-intel) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: > I agree that we should have a more granular set of headers in SYCL. I think the right way to do this is to keep `<sycl/sycl.hpp>` as the "uber" header that includes everything, and then we can also create a set of more granular headers that include specific features. We can also create separate headers for each extension (to the extent possible) as part of this effort.

And I wasn't talking about `sycl/sycl.hpp`, it was about `sycl/detail/core.hpp` (see https://github.com/intel/llvm/pull/12890 that introduced it and there was a bunch of changes after). The problem we have is that it's way too easy to add all the new stuff in such a way that `sycl::handler`/`sycl::queue` depend on it unconditionally and then there is no way to make *any* meaningful separation of features.

Comment (lbushi25) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: The relevant include chain is core.hpp includes queue.hpp includes info_desc_helpers.hpp includes info_desc.hpp includes forward_progress.hpp.
I'll see tomorrow if I can perhaps skip the inclusion of forward_progress.hpp from info_desc.hpp without deviating too much from our usual code structure.

Comment (aelovikov-intel) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: It's not as much about that header but more about all the templated code that FE has to parse that you put into other headers.

Comment (lbushi25) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: I see what you mean. As a worst case scenario, from a client's perspective, a user that just wants to perform basic matrix operations using SYCL does not need anything related to this extension and yet by including sycl/sycl.hpp they will incur some of the compile-time cost of the template parsing that happens even when the templates are not instantiated. 

As @gmlueck said above, it would be ideal to separate headers for big extensions but that could be hard to do. The bigger an extension, the more coupled it will be with existing code. For example, due to a stipulation in the spec of this extension about verifying the properties required by a kernel, I had to go and augment the implementation of the already existing function processProperties in handler.hpp. The alternative, in such cases, would be to add preprocessor conditionals enabling/disabling compilation of extension code.

Comment (gmlueck) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: > For example, due to a stipulation in the spec of this extension about verifying the properties required by a kernel, I had to go and augment the implementation of the already existing function processProperties in handler.hpp.

Do you mean this part of the spec:

> If a kernel is submitted to a device that cannot satisfy the request for specific progress guarantees, the implementation must throw an `exception` with the `errc::feature_not_supported` error code.

Properties are getting integrated into a lot of APIs, so I think many headers will end up needing to include the base definitions for properties.

> The alternative, in such cases, would be to add preprocessor conditionals enabling/disabling compilation of extension code but this can quickly become unwieldy.

I'm not sure this is a good path.  We previously considered adding a command line switch that would enable/disable each property, but we rejected this because it introduced too many testing combinations.

Comment (lbushi25) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: I agree. I do think that its better to continue doing what we've been doing and perhaps slowly start to separate headers for extensions where it is possible to do so.

Comment (lbushi25) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: Tagging @AlexeySachkov for input on this discussion.

Comment (lbushi25) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: @aelovikov-intel Do we have your approval on this?
Otherwise, we can put this on hold while an outline of a standard approach on how to deal with these concerns is documented somewhere. I don't want to resort to ad-hoc approaches.

Comment (steffenlarsen) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: Please correct me if I am wrong, @aelovikov-intel, but your point here is not that including into `core.hpp` affects users, but rather that we keep `core.hpp` as bare-bones as possible to help reduce compilation overhead in our tests, as the compiler has to re-parse our headers for even small tests and the less overhead we can have the better.

From a user's perspective, all they know exists from the SYCL spec is `<sycl/sycl.hpp>` and we make no exception for extensions currently. I would personally love to see that change in the future, but for now `sycl.hpp` must include everything and `core.hpp` should include only the essentials.

Comment (aelovikov-intel) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: > Please correct me if I am wrong, @aelovikov-intel, but your point here is not that including into core.hpp affects users, but rather that we keep core.hpp as bare-bones as possible to help reduce compilation overhead in our tests

Kind of, see below.

> Perhaps this points to a larger systemic issue with the way we've designed our tests/code but that will need to be handled separately in my opinion.

Yes, there is one, but we *do* work on it. https://github.com/intel/llvm/pull/12890 introduced `<sycl/detail/core.hpp>` and later https://github.com/intel/llvm/pull/13210 introduced an in-tree test supposed to keep the growth of the newly introduced header file under control. As such, saying 

> For now, however, I'm just respecting the status quo as previous extensions have done.

and implying that nothing is getting fixed outside of this PR is incorrect. Note, that if would be impossible for 1-2 people to try to gradually improve things not working full time on that task if other dozens of contributors would just keep bloating `handler.hpp`/`queue.hpp` (that are part of `detail/core.hpp` for obvious reasons) as being done in this PR.

Now, switching to what can be done here. I think you can use some kind of visitor pattern. `<oneapi_properties>::has_property` has to have a loop inside its implementation. What you can do is this:

```
handler.hpp:
// SFINAE-visitor
template <Property>
struct ProcessForwardProgressProperty {
  void operator() {
    // do nothing for non-forward-progress property
  }
};
...
 loop over all the properties passed to the method and call the above `operator()` with it.

forward_progress_property:
// define property and provide partial specialization for the above helper.
```

Comment (lbushi25) [reply sycl/test/include_deps/sycl_accessor.hpp.cpp:79]: @aelovikov-intel 
I discussed this with @AlexeySachkov and we decided to open an issue regarding this: https://github.com/intel/llvm/issues/13774. I've assigned myself to it and will start working on it immediately.
In the meantime, let's proceed with getting this PR merged.

**Comment by lbushi25** [comment]
This has introduced post-commit errors which are being addressed by https://github.com/intel/llvm/pull/13793/

**Pull request merged**
